### ISSUE TITLE
Amplify.configure(...) throws AmplifyException instead of AmplifyAlreadyConfigured

### ISSUE THREAD

**Issue by fpabl0** [aws-amplify/amplify-flutter#412]
**Describe the bug**
[AmplifyDocs](https://docs.amplify.aws/start/getting-started/integrate/q/integration/flutter#initializing-the-amplify-flutter-library) says we should not reconfigured Amplify. Doing it will result in a `AmplifyAlreadyConfiguredException`:

```dart
  try {
    await Amplify.configure(amplifyconfig);
  } on AmplifyAlreadyConfiguredException {
    print("Amplify was already configured. Was the app restarted?");
  }
```

However, `Amplify.configure(...)` does not specifically throw `AmplifyAlreadyConfiguredException`, it just throws an `AmplifyException` with `AmplifyAlreadyConfigured` as the underlying exception.

**To Reproduce**
1. Run the example in [docs](https://docs.amplify.aws/start/getting-started/integrate/q/integration/flutter#initializing-the-amplify-flutter-library).
2. After app is running successfully, perform a Restart.
3. See the error.

**Expected behavior**
Instead of a `AmplifyException`, I expect a `AmplifyAlreadyConfiguredException`.

**Screenshots**
<img width="707" alt="Captura de Pantalla 2021-02-27 a la(s) 10 53 14" src="https://user-images.githubusercontent.com/12239342/109392511-68bf9b00-78ea-11eb-87c5-697748afc7a6.png">

https://user-images.githubusercontent.com/12239342/109392677-788baf00-78eb-11eb-830f-a0da94de09f9.mp4

**Platform**
Amplify Flutter current supports iOS and Android.  This issue is reproducible in (check all that apply):
[] Android
[x] iOS (I have only tried iOS simulator)

**Smartphone Simulator (please complete the following information):**
 - Device: iPhone SE (2nd generation)
 - OS: iOS 13.5
 - Simulator Version:  11.5 (921.9.1)

**Comment by jcowley** [comment]
What would be even more useful would be the ability to detect if Amplify has been configured without having to deal with a thrown exception in the first place. This exception comes up in the debugger every time you reload the app, regardless of whether you catch and handle it a high level in your code. So, this is a major annoyance for developers.

**Comment by brunovsiqueira** [comment]
Same here.

**Comment by raywhiteside** [comment]
Earlier topic:

https://github.com/aws-amplify/amplify-flutter/issues/124#event-4186447787

I'm not sure why a specific exception avoids this issue. Is the expectation that developers catch this exception in development to support hot reload, but avoid catching it in production? Why is this better than being able to check the state before initializing, so the same code supports both cases?

**Comment by A-Kurimoto** [comment]
I avoided this error by following code.
Please improve!

```
 try {
    await Amplify.configure(Environment.amplifyConfig.value);
  } on AmplifyAlreadyConfiguredException {
    print(
        "Tried to reconfigure Amplify; this can occur when your app restarts on OS.");
  } on AmplifyException catch (e) {
    if (e.underlyingException
        .contains('Amplify has already been configured.')) {
      print('ignore');
    } else {
      throw e;
    }
  }
```

**Comment by jcowley** [comment]
@A-Kurimoto: Doesn't really solve the issue for devs. It still throws an exception every time you reload the app when running in debug mode, which you have to step over in the debugger.

Looking at the history of this, it looks like an `isConfigured` method was added at one point (#303). Then after "internal discussion" it was determined that throwing an exception was better and it was reverted (#316). 

It's not clear why throwing an exception would be considered a better solution than simply being able to check to see if the library has been configured.

**Comment by Amplifiyer** [comment]
@jcowley, @raywhiteside thank you for your feedback. When we made the decision to revert `isConfigured`, we weren't aware of this annoyance with debugger and debugging apps. I agree that it's not a good developer experience and so we will be putting the `isConfigured` back. It will be part of our next release.

**Comment by jcowley** [comment]
@Amplifiyer: Thank you!

**Comment by brunovsiqueira** [comment]
any updates on this?

**Comment by infernal008** [comment]
Hello @Amplifiyer, I would just like to confirm if `isConfigured` is still a work in progress or is suppossedly be done as of this time?

I am able to call isConfigured but have the following observations:

1. I tried printing the value of isConfigured before and after calling Amplify.configure()
```
print(Amplify.isConfigured); // false
Amplify.configure(config);
print(Amplify.isConfigured); // true
```
2. You can see from above that the value of isConfigured changes to true after - which is an expected behavior.
3. Doing hot-reload and having to print the isConfigured value again returns `false`.
4. From my understanding, the isConfigured value should be presisted even after hot-reload, right?

I'm not sure if the release mentioned earlier is what is available right now or not, I just happen to be able to call the discussed isConfigured attribute as of this time and would like to share my observations.

**Comment by jcowley** [comment]
@infernal008: Doesn't work for me either. Hot reload appears to clear `isConfigured` but not the actual configuration. So, still getting the exception on every hot reload.

**Comment by infernal008** [comment]
@jcowley, what I have done so I won't get stuck in it during development is I just disabled the _somehow_ aggressive exception breakpoints in vscode. 

Doing hot reload seems to work fine and have no problems using the amplify plugins - atleast I can move on and could just come back at it when I'm done with other stuff. I aslo don't currently have a mechanism to check if amplify is already configured or not but it just works. Though I wouldn't hold much into it as it is not really a concrete solution/fix.

**Comment by haverchuck** [comment]
@infernal008 @fpabl0 @brunovsiqueira @A-Kurimoto @raywhiteside 

We have just released version 0.1.4 and have addressed this. The isConfigured getter will not work immediately on hot restart (because hot restart destroys dart state), but if you call configure on a hot restart just as you would on an initial app start, we detect in our platform-specific code whether amplify has already been configured, swallow the resulting error that is sent to the dart code via the method channel, re-wire the amplify plugins in dart, and reset the isConfigured value.

tldr: With version 0.1.4 you can treat a hot restart exactly as you would an initial app start, and the behavior (in terms of dart state) should be identical.

**Comment by infernal008** [comment]
> @infernal008 @fpabl0 @brunovsiqueira @A-Kurimoto @raywhiteside
> 
> We have just released version 0.1.4 and have addressed this. The isConfigured getter will not work immediately on hot restart (because hot restart destroys dart state), but if you call configure on a hot restart just as you would on an initial app start, we detect in our platform-specific code whether amplify has already been configured, swallow the resulting error that is sent to the dart code via the method channel, re-wire the amplify plugins in dart, and reset the isConfigured value.
> 
> tldr: With version 0.1.4 you can treat a hot restart exactly as you would an initial app start, and the behavior (in terms of dart state) should be identical.

Hello @haverchuck, thanks for the tag with your update. I'm currently on 0.1.4 and it indeed seems to work as intended - no exceptions/messages are thrown regarding reconfiguration during hot restart. This is great for me as it seems to be smoother now in terms of using the plugins - it just works.

**Comment by haverchuck** [comment]
Closing this issue based on feedback from @infernal008 .
